# DS 2.2 Deep Learning: 
- **Make School Courses repo**: click [here](https://github.com/Make-School-Courses/DS-2.2-Deep-Learning)
- **My GitHub Repo**: click [here](https://github.com/SamuelFolledo/DS2.2-Deep-Learning)


**Disclaimer:** Table of Contents only works on Jupyter Notebook

## Table of Contents
1. **[Introduction to Deep Learning](#day1)**
2. **[What is Neural Network](#day2)**
3. **[Neural Network from Scratch](#day3)**
4. **[Intro To Keras](#day4)**
5. **[Backward Propagation](#day5)** - 11/9/2020
6. **[](#day6)**
7. **[](#day7)**
8. **[Convolutional Neural Network](#day8)** - 11/18/2020
    - [Lab 1](#lab1)
9. **[Quiz and Lab day](#day9)**
10. **[Recurrent Neural Network](#day10)** 11/30/2020
11. **[](#day11)**
12. **[](#day12)**
13. **[](#day13)**
14. **[](#day14)**

<a name="day1"></a> 
# Day 1: Introduction to Deep Learning
https://docs.google.com/presentation/d/1vxQ_R8-gEtR896EoC1_ObSKoWJZje40H8wcXzdltjpo/edit#slide=id.g9e8c78a5a9_0_18

### What We’re Going to Learn
- High Level Overview of Deep Learning
- Applications of Deep Learning 
- What is a Neural Network


## Deep Learning
- Deep learning is a **subfield of machine learning**
- Deep learning is **all about neural networks**
- Deep learning **refers to large neural networks**
- Neural networks have been around for a while but is only relatively recently that we have the computing power to train and run the these larger more complex neural networks
- The deep part comes from **having many layers in the networks**

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day1-1.png?raw=true" width="700">

### Why Deep Learning?
*“The analogy to deep learning is that the rocket engine is the deep learning models and the fuel is the huge amounts of data we can feed to these algorithms.”* - Andrew Ng

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day1-2.png?raw=true" width="500">

### Drawbacks of Deep Learning
- Interpretability is harder, very black box 
- Doesn’t work well with small datasets
    - the less data you have, the worse it's going to perform
    - more instances/samples (~10,000) not how many features
- Needs lots of computing power

## Neural Networks
- Modeled loosely on the human brain
- Composed of many simple processing nodes or “neurons”
<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day1-3.png?raw=true" width="700">

## A Neuron
- For most modern deep learning algorithms we use what is called a “sigmoid neuron”
- In order to understand these better let’s take a look at the idea they were based on, a perceptron, one of the first types of artificial neurons

## Perceptrons
- A perceptron takes in multiple inputs and produces a single binary output
- Each of these inputs x have a weight w, what would weights do?
	◦ Weights are used to add importance of a feature, giving it more influence on the output
<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day1-4.png?raw=true" width="700">


## Deep Neural Networks
We add all those x*w’s together to get a sum
<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day1-5.png?raw=true" width="500">


## Activation Function //58m
How do we know whether to give a 0 or 1, that’s the activation functions job! One example is the Unit Step Activation Function
<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day1-6.png?raw=true" width="700">


### What are some drawbacks of this decision making model? //1h1m
- Setting an absolute threshold is very not a good decision


## Sigmoid Neuron: Much Less Harsh //1h4m
<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day1-7.png?raw=true" width="700">

## Sigmoid Neuron Output //1h5

- Our output is no longer 0 or 1
- It is now a value from 0 - 1 and can be interpreted like a probability
- We now have more control over how that output is interpreted

## Putting it all together
- A neural network is really just a bunch of interconnected neurons, the output from one layer or neurons can then be used as one the inputs for the next layer
- Through a series of layers different patterns of neurons are “activated” producing different end results for different inputs

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day1-8.png?raw=true" width="700">

## Next Steps
- Math happening behind the scenes?
- What about those algorithms?
- How does it actually get the inputs (features)?
- What sorts of parameters and choices can we mess with, how do we optimize?
- How do we choose how many layers and neurons?
- What tools are available to us?
- How can we apply this to a real world problem?

<a name="day2"></a> 
# Day 2 What is Neural Network?
https://github.com/Make-School-Courses/DS-2.2-Deep-Learning/blob/master/Lessons/WhatisNeuralNetwork.md

### Learning Objectives
	•	What is Neural network
	•	We learn Forward and backward propagation in NN
	•	We implement a NN entirely in numpy

## What is Neural Network?
It is a **computational system inspired by the Structure, Processing Method and Learning Ability similar to our biological brain**

### Characteristics of Artificial Neural Networks
	•	A large number of very simple processing neuron-like processing elements
	•	A large number of weighted connections between the elements
	•	Distributed representation of knowledge over the connections
	•	Knowledge is acquired by network through a learning process

## What is perceptron?
	•	A perceptron can be understood as anything that **takes multiple inputs and produces one output**

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day2-1.png?raw=true" width="700">


## Multi-layer perceptron (MLP)
	•	MLP is the stack of perceptrons
    
<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day2-2.png?raw=true" width="700">

	•	In this image, the yellow nodes are inputs, the blue nodes (at each vertical) are hidden layers and the orange ones are output of the MLP

## Forward and backward propagation
NN **takes several input, processes it through multiple neurons from multiple hidden layers and returns the result using an output layer.** This result estimation process is technically known as **“Forward Propagation“**

Next, we **compare the result with actual output**. The task is to make the output to neural network as close to actual (desired) output. This defines our cost function.
We try to **obtain the weight of neurons such that the NN total error (our cost function) being minimized**. This process is known as **“Backward Propagation“**

In [3]:
import numpy as np

#creating a vector
vector = np.array([5,2,1])
print(type(vector))

<class 'numpy.ndarray'>


In [4]:
import numpy as np

# check this out:
# https://www.analyticsvidhya.com/blog/2017/05/neural-network-from-scratch-in-python-and-r/
# Input array
X=np.array([[1,0,1,0],[1,0,1,1],[0,1,0,1]])

#Output
y=np.array([[1],[1],[0]])


#Sigmoid Function //gives us a value between 0 and 1 or Probability
def sigmoid (x):
    return 1/(1 + np.exp(-x))


#Derivative of Sigmoid Function
def derivatives_sigmoid(x):
    return x * (1 - x)



#Variable initialization
epoch=5000 #Setting training iterations
lr=0.1 #Setting learning rate
inputlayer_neurons = X.shape[1] #number of features in data set
hiddenlayer_neurons = 3 #number of hidden layers neurons
output_neurons = 1 #number of neurons at output layer

#weight and bias initialization
wh=np.random.uniform(size=(inputlayer_neurons,hiddenlayer_neurons))
bh=np.random.uniform(size=(1,hiddenlayer_neurons))
wout=np.random.uniform(size=(hiddenlayer_neurons,output_neurons))
bout=np.random.uniform(size=(1,output_neurons))


for i in range(epoch):
    #Forward Propogation
    hidden_layer_input1=np.dot(X,wh)
    hidden_layer_input=hidden_layer_input1 + bh
    hiddenlayer_activations = sigmoid(hidden_layer_input)
    output_layer_input1=np.dot(hiddenlayer_activations,wout)
    output_layer_input= output_layer_input1+ bout
    output = sigmoid(output_layer_input)

    #Backpropagation
    D = y-output
    slope_output_layer = derivatives_sigmoid(output)
    slope_hidden_layer = derivatives_sigmoid(hiddenlayer_activations)
    d_output = D * slope_output_layer
    Error_at_hidden_layer = d_output.dot(wout.T)
    d_hiddenlayer = Error_at_hidden_layer * slope_hidden_layer
    wout += hiddenlayer_activations.T.dot(d_output) *lr
    bout += np.sum(d_output, axis=0,keepdims=True) *lr
    wh += X.T.dot(d_hiddenlayer) *lr
    bh += np.sum(d_hiddenlayer, axis=0,keepdims=True) *lr

print(output)

[[0.98328823]
 [0.97397276]
 [0.0367804 ]]


## How do we update the weights to minimize the error?
1. First we should define the cost function. for our example here the MSE is our cost function:

$E= \frac{1}{2} ({\bf y}_t - {\bf y}_p)^T ({\bf y}_t - {\bf y}_p)$

2. We update the weight (${\bf W}_i$ and ${\bf W}_h$) such that the error, $E$, being minimized. The most popular algorithm is Gradient Descent:

${\bf W}_h = {\bf W}_h + \eta {\partial E}/{\partial {\bf W}_h} $

For our above example we can show that:

${\partial E}/{\partial {\bf W}_h} = ({\bf y}_t - {\bf y}_p) {\bf y}_p (1 - {\bf y}_p)\bf {h}$

where ${\bf h} = \sigma({\bf W}_i {\bf x}_i + {\bf b}_i)$

In above code:
$D = {\bf y}_t - {\bf y}_p$

${\bf y}_p (1 - {\bf y}_p)$ = slope_hidden_layer

$\bf {h}$ = hiddenlayer_activations

## Other Resources:
- https://enlight.nyc/projects/neural-network/
- WTF is Tensor: https://www.kdnuggets.com/2018/05/wtf-tensor.html


<a name="day3"></a> 
# Day 3: Neural Network from Scratch
https://docs.google.com/presentation/d/1QstriVSuk8ZXWYjPrJ6SZUTjjYTC9Z6HHwJAun-hBEE/edit#slide=id.p

### Interview Questions
- What is the bias variance tradeoff in machine learning?
    - Bias is the simplifying assumptions made by the model to make the target function easier to approximate. Variance is the amount that the estimate of the target function will change given different training data. Trade-off is tension between the error introduced by the bias and the variance
    - https://machinelearningmastery.com/gentle-introduction-to-the-bias-variance-trade-off-in-machine-learning/#:~:text=Bias%20is%20the%20simplifying%20assumptions,the%20bias%20and%20the%20variance.

- What is an example of a low bias ML algorithm?
    - Low Variance: Suggests small changes to the estimate of the target function with changes to the training dataset.
        - Examples of low-variance machine learning algorithms include: Linear Regression, Linear Discriminant Analysis and Logistic Regression.
    - High Variance: Suggests large changes to the estimate of the target function with changes to the training dataset.
        - Examples of high-variance machine learning algorithms include: Decision Trees, k-Nearest Neighbors and Support Vector Machines.
        - Can potentially overfit

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day3-1.png?raw=true" width="750">

- Bias
    - Shifts curve of our activation function
    - Weights can help give a guidance towards your model
    - Help optimize parameters to help
    - Allows us to place significance over selected parameters
    - Traditionally set to 1

## Tutorial: Machine Learning for Beginners: An Introduction to Neural Networks
https://victorzhou.com/blog/intro-to-neural-networks/

## Neuron
- A neuron takes inputs, does some math with them, and produces one output

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day3-2.png?raw=true" width="500">

1. First, each input is multiplied by a weight:

    `x1 -> x1 * w1`
    `x2 -> x2 * w2`
    
2. All the weighted inputs are added together with a bias *b*:
    
    `(x1 * w1) + (x2 * w2) + b`
    
3. The sum is passed through an activation function:

    `y = f(x1*w1 + x2*w2 + b)`

### Activation Function
- used to turn an unbounded input into an output that has a nice, predictable form. 
- A commonly used activation function is the sigmoid function

## Sigmoid
- The sigmoid function only outputs numbers in the range (0,1). You can think of it as compressing (−∞,+∞) to (0,1) - big negative numbers become ~0, and big positive numbers become ~1.

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day3-3.png?raw=true" width="500">

### Neuron Code in Python

In [5]:
import numpy as np

def sigmoid(x):
    # Our activation function: f(x) = 1 / (1 + e^(-x))
    return 1 / (1 + np.exp(-x))

class Neuron:
    """This class represents an articfical sigmoid neuron"""

    def __init__(self, weights, bias, activation_function):
        """ initializes this neuron with weights, inputs, and bias"""
        self.weights = weights
        self.bias = bias
        self.activation_function = activation_function

    def feedForward(self, inputs):
        """steps to return the y output using dot function"""
        result = np.dot(self.weights, inputs) + self.bias
        return self.activation_function(result)

In [6]:
weights = np.array([0,1])
bias = 1
inputs = np.array([2,4])
#instantiate a neuron
neuron = Neuron(weights, bias, sigmoid)
print(neuron.feedForward(inputs))

0.9933071490757153


## Simple Neural Network

### The Neural Network's FeedForward
<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day3-4.png?raw=true" width="500">

In [7]:
class SimpleNeuralNetwork:
    
    def __init__(self, weights, bias):
        self.weights = weights
        self.bias = bias
        
        #inputs or X's
        self.inputs = inputs
        
        #hidden layer
        self.h1 = Neuron(weights, bias, sigmoid)
        self.h2 = Neuron(weights, bias, sigmoid)
        
        #output layer
        self.o1 = Neuron(weights, bias, sigmoid)
        
    def feedForward(self, inputs):
        """get hidden layer outputs"""
        out_h1 = self.h1.feedForward(inputs)
        out_h2 = self.h2.feedForward(inputs)
        
        out_o1 = self.o1.feedForward(np.array([out_h1, out_h2]))
        return out_o1

In [8]:
weights = np.array([0, 1])
bias = 1

simpleNN = SimpleNeuralNetwork(weights, bias)
inputs = np.array([2, 4])

print(simpleNN.feedForward(inputs))

0.8800925786929503


## HW: 
Check out this [tutorial](https://victorzhou.com/blog/intro-to-neural-networks/) to prep for the backpropagation step

In [9]:
myList = [1,2,3,4,5]
myList[-1]

5

<a name="day4"></a> 
# Day 4: Intro to Keras
https://docs.google.com/presentation/d/1QstriVSuk8ZXWYjPrJ6SZUTjjYTC9Z6HHwJAun-hBEE/edit#slide=id.p

## Back propagation
Once we have randomly initialized the weights how do we tune them?
- We need to be able to measure how “good” our neural network is doing often referred to as a loss function
- Once we find a measure we need to adjust the weights in a way that will minimize the “cost” of this measure
- Trying to calculate every combination of weights to see which one reduces the cost the most is not efficient
- To minimize the loss function a technique called gradient descent is often used in ML

## The Loss Function
- *“A loss function or cost function is a function that maps an event or values of one or more variables onto a real number intuitively representing some "cost" associated with the event”*
- There are many different loss functions, the one we will use in the example is called Mean Squared Error (MSE)

## Mean Squared Error
- n: number of samples     
- y pred: output being predicted by our model
- y true: actual output from training data 

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day4-1.png?raw=true" width="500">

We square the error and take the average because it makes [the math nicer](https://www.benkuhn.net/squared/)


## Gradient Descent
- How loss function minimize the loss or "train" our network

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day4-2.png?raw=true" width="750">

## Notes
- Bad thing about `train-test-split` we can lose valuable information by not including data in the test
    - Not having certain valuable information can make things unpredictable
- `Cross validation` validate data by the fold, then the other fold, and then avarage the result to get a balance training and result

In [10]:
from sklearn import datasets
from sklearn.model_selection import train_test_split

iris = datasets.load_iris()
X, y = iris.data, iris.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

### Category vs. Numerical
- **Keras takes numerical, not categorical**
- Category removes the greater value and evens the playing field for different possible targets

In [11]:
from keras.utils import np_utils

y_train_one_hot = np_utils.to_categorical(y_train)
y_test_one_hot = np_utils.to_categorical(y_test)

# print(y_test_one_hot)

## Deep Learning in Keras

### Deep Learning Platforms in Python
1. Keras 
2. Tensorflow 
3. Pytorch 
4. Caffe 
5. Theano 
6. CNTK 7- MXNET

### Why we use Keras in DS 2.2 ?
- A focus on user experience, easy to build and train a deep learning model
- Easy to learn and easy to use
- Large adoption in the industry and research community
- Multi-backend, multi-platform
- Easy productization of models

## Keras has two API Styles

### 1. Sequential API
- Dead simple
- **Only for single-input, single-output, sequential layer stacks**
- Good for **70+%** of use cases

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day4-3.png?raw=true" width="1000">

---

### 2. Functional API
- Like playing with Lego bricks
- **Multi-input, multi-output, arbitrary static graph topologies**
- Good for **95%** of use cases

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day4-4.png?raw=true" width="800">

### Jess's example

In [12]:
from keras.models import Sequential
from keras.layers.core import Dense, Activation

model = Sequential()
model.add(Dense(16, input_shape=(4,))) #put its input. What is the output shape of that layer going to be?
model.add(Activation('sigmoid'))#specified activation function for those layers
model.add(Dense(3)) #lessen the nodes
model.add(Activation('sigmoid'))

model.compile(optimizer='SGD', loss='categorical_crossentropy', metrics=['accuracy'])

# Train
model.fit(X_train, y_train_one_hot, epochs=100, batch_size=1) #how many time do we descend in our Gradient Decend; batch_size = how many samples considered each time we update our weights

Epoch 1/100
105/105 [==============================] - 0s 892us/step - loss: 1.1350 - accuracy: 0.2857
Epoch 2/100
105/105 [==============================] - 0s 1000us/step - loss: 1.1085 - accuracy: 0.2857
Epoch 3/100
105/105 [==============================] - 0s 932us/step - loss: 1.0905 - accuracy: 0.3429
Epoch 4/100
105/105 [==============================] - 0s 826us/step - loss: 1.0761 - accuracy: 0.3524
Epoch 5/100
105/105 [==============================] - 0s 857us/step - loss: 1.0617 - accuracy: 0.4000
Epoch 6/100
105/105 [==============================] - 0s 911us/step - loss: 1.0461 - accuracy: 0.2286
Epoch 7/100
105/105 [==============================] - 0s 782us/step - loss: 1.0302 - accuracy: 0.4286
Epoch 8/100
105/105 [==============================] - 0s 893us/step - loss: 1.0137 - accuracy: 0.3238
Epoch 9/100
105/105 [==============================] - 0s 834us/step - loss: 0.9959 - accuracy: 0.4476
Epoch 10/100
105/105 [==============================] - 0s 771us/step - 

In [13]:
loss, accuracy = model.evaluate(X_test, y_test_one_hot)
print(accuracy)

2/2 [==============================] - 0s 898us/step - loss: 0.2084 - accuracy: 0.8889
0.8888888955116272


## Activity: Apply NN with Keras on iris data
- Use Sequential API for Keras
- Use 70 percent of data for train
- Use one-hot encoding for labels with from keras.utils import np_utils
- Define two layers fully connected network
- Define categorical_crossentropy as the loss (cost) function

In [14]:
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.utils import np_utils
from sklearn.preprocessing import LabelEncoder
from sklearn import datasets
from sklearn.model_selection import train_test_split

iris = datasets.load_iris()
X, y = iris.data, iris.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

y_train_one_hot = np_utils.to_categorical(y_train)
y_test_one_hot = np_utils.to_categorical(y_test)

# print(y_one_hot)

model = Sequential()
model.add(Dense(16, input_shape=(4,)))
model.add(Activation('sigmoid'))
model.add(Dense(3))
model.add(Activation('softmax'))
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=["accuracy"])
model.fit(X_train, y_train_one_hot, epochs=100, batch_size=1, verbose=0);
loss, accuracy = model.evaluate(X_test, y_test_one_hot, verbose=0)
print("Accuracy = {:.2f}".format(accuracy))

Accuracy = 0.98


## Activity: Apply NN with Keras on iris data with Functional API

In [15]:
from keras.layers import Input, Dense
from keras.models import Model
from keras.utils import np_utils
from sklearn.preprocessing import LabelEncoder
from sklearn import datasets
from sklearn.model_selection import train_test_split

iris = datasets.load_iris()
X, y = iris.data, iris.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

y_train_one_hot = np_utils.to_categorical(y_train)
y_test_one_hot = np_utils.to_categorical(y_test)

# print(y_one_hot)

inp = Input(shape=(4,))
x = Dense(16, activation='sigmoid')(inp)
out = Dense(3, activation='softmax')(x)
model = Model(inputs=inp, outputs= out)
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=["accuracy"])
model.fit(X_train, y_train_one_hot, epochs=100, batch_size=1, verbose=0);
loss, accuracy = model.evaluate(X_test, y_test_one_hot, verbose=0)
print("Accuracy = {:.2f}".format(accuracy))

Accuracy = 0.98


<a name="day5"></a> 
# Day 5: Backward Propagation
https://docs.google.com/presentation/d/1QstriVSuk8ZXWYjPrJ6SZUTjjYTC9Z6HHwJAun-hBEE/edit#slide=id.g9fd5bc0407_0_5

## Back Propagation
- Once we have randomly initialized the weights how do we tune them?
- We need to be able to **measure how “good” our neural network** is doing often referred to as a **loss function**
- Once we find a measure we **need to adjust the weights in a way that will minimize the “cost”** of this measure
- Trying to calculate every combination of weights to see which one reduces the cost the most is not efficient
- To **minimize the loss function a technique called gradient descent** is often used in ML

## The Loss Function
- “A loss function or cost function is a **function that maps an event or values of one or more variables onto a real number intuitively representing some "cost" associated with the event”**
- There are many different loss functions, the one we will use in the example is called **Mean Squared Error (MSE)**

## Back Propagation
- We are going to work through [this article](https://towardsdatascience.com/backpropagation-for-people-who-are-afraid-of-math-936a2cbebed7)

#### Backpropagation Steps:
1. Calculate the cost function, C(w)
2. Calculate the gradient of C(w) with respect to (w.r.t) all the weights, w, and biases, b, in your neural network (NN)
3. Adjust the w and b proportionally to their gradients
Also check out [this article](https://towardsdatascience.com/the-maths-behind-back-propagation-cf6714736abf) for more math 


In [17]:
import numpy as np

class Neuron:
  """This class represents an articfical sigmoid neuron"""

  def __init__(self, weights, bias, activation_function):
    """ initializes this neuron with weights, inputs, and bias"""

    self.weights = weights
    self.bias = bias
    self.activation_function = activation_function

  def feedforward(self, inputs):
    #TODO: code the steps to return the y output
    result = np.dot(self.weights, inputs) + self.bias
    return self.activation_function(result)

weights = np.array([0,1])
bias = 1
inputs = np.array([2,4])
#instantiate a neuron
neuron = Neuron(weights, bias, sigmoid)
#print(neuron.feedforward(inputs))

In [18]:
class SimpleNeuralNetwork:

  def __init__(self, weights, bias):
    self.weights = weights
    self.bias = bias

    #hidden layer
    self.h1 = Neuron(weights, bias, sigmoid)
    self.h2 = Neuron(weights, bias, sigmoid)

    #output layer
    self.o1 = Neuron(weights, bias, sigmoid)

  def feedforward(self, inputs):
    #get hidden layer outputs
    out_h1 = self.h1.feedforward(inputs)
    out_h2 = self.h2.feedforward(inputs)

    out_o1 = self.o1.feedforward(np.array([out_h1, out_h2]))

    return out_o1

#y_pred and y_true are numpy arrays
def mse_loss(y_true, y_pred):
  return ((y_true - y_pred) ** 2).mean()
  
weights = np.array([0,1])
bias = 1
simplenn = SimpleNeuralNetwork(weights, bias)
inputs = np.array([2,4])
print(simplenn.feedforward(inputs))

y_true = np.array([1,0,0,1])
y_pred = np.array([0,0,0,0])
print(mse_loss(y_true, y_pred))

0.8800925786929503
0.5


### Derivatives Rules
https://www.mathsisfun.com/calculus/derivatives-rules.html

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day5-1.png?raw=true" width="500">

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day5-2.png?raw=true" width="500">

## Gradient Descent
https://www.khanacademy.org/math/multivariable-calculus/multivariable-derivatives/gradient-and-directional-derivatives/v/gradient

<img src="https://github.com/SamuelFolledo/DS2.2-Deep-Learning/blob/master/static/screenshots/day5-3.png?raw=true" width="750">

# Important Notebooks
- [Deep Learning Glossary](https://github.com/Make-School-Courses/DS-2.2-Deep-Learning/blob/master/Lessons/DeepLearningGlossary.md)

<a name="day8"></a> 
# Day 8: Convolutional Neural Network
- **Sequential** is a NN model that allows us to add layer by building it step by step
- **Dense** adds a layer of neurons


<a name="lab1"></a> 
### Lab 1

Tutorial: [Python Convolutional Neural Network: Creating a CNN in Keras, TensorFlow and Plain Python](https://missinglink.ai/guides/convolutional-neural-networks/python-convolutional-neural-network-creating-cnn-keras-tensorflow-plain-python/)


In [23]:
from keras.models import Sequential
from keras.layers import Dense, Flatten, Activation
from keras.layers import Conv2D
from keras.utils.np_utils import to_categorical

#TODO: get train and test data
from keras.datasets import mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train = X_train.reshape(60000,28,28,1)
X_test = X_test.reshape(10000,28,28,1)
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

#create model
model = Sequential()

#add convolutional layer
filters = 64
kernel_size = 3 #3x3 matrix
activation_function = "relu" #rectified linear activation function
model.add(Conv2D(filters, kernel_size=kernel_size, activation=activation_function, input_shape=(28,28,1)))

#TODO: might add some other Dense layers here and Flatten()
model.add(Conv2D(32, kernel_size=3, activation=activation_function))
model.add(Flatten())
model.add(Dense(10, activation="softmax"))

#compile the model by setting the optimizer, loss function and metrics
model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])

# model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

#fit to train and test
# model.fit(X_train, y_train, epochs=3)
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=3)

#predict
# model.predict(X_test)
model.predict(X_test[:4])

Epoch 1/3
1875/1875 [==============================] - 73s 39ms/step - loss: 0.2210 - accuracy: 0.9521 - val_loss: 0.0966 - val_accuracy: 0.9716
Epoch 2/3
1875/1875 [==============================] - 78s 42ms/step - loss: 0.0660 - accuracy: 0.9801 - val_loss: 0.0848 - val_accuracy: 0.9750
Epoch 3/3
1875/1875 [==============================] - 81s 43ms/step - loss: 0.0479 - accuracy: 0.9848 - val_loss: 0.0758 - val_accuracy: 0.9794


array([[4.77071724e-14, 4.02404522e-16, 1.33951239e-10, 2.40395437e-08,
        6.78720036e-16, 3.99273408e-14, 1.53550912e-18, 1.00000000e+00,
        1.03526963e-12, 9.96556726e-10],
       [8.09913914e-09, 3.13066617e-09, 9.99999762e-01, 4.81141849e-10,
        3.86484396e-14, 2.07901814e-14, 2.24898642e-08, 3.40396873e-13,
        2.81714961e-07, 2.78229761e-15],
       [4.30938529e-09, 9.99942541e-01, 1.08588370e-06, 4.05593141e-08,
        4.64521044e-08, 4.60761065e-08, 3.15033262e-08, 1.80602513e-07,
        5.60371482e-05, 5.86901416e-09],
       [9.99990940e-01, 2.75990428e-12, 1.09197366e-07, 7.00521509e-13,
        2.02911611e-11, 9.87869253e-09, 3.80378606e-06, 6.28115338e-11,
        1.14980672e-07, 5.02158673e-06]], dtype=float32)

In [25]:
y_test[:4]

array([[0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], dtype=float32)

## Milad's Convolutional Neural Network Notebook
- [Source](https://github.com/Make-School-Courses/DS-2.2-Deep-Learning/blob/master/Lessons/ConvolutionalNeuralNetwork.md)

### Learning Objectives
- we learn what are Convolutional Neural Network (CNN)
- Will talk about CNN components such as stride, max or average pooling
- Discuss how we can obtain the parameters for CNN

### Convolutional Neural Network (CNN)
- CNN is basically two dimensional configuration of neural networks
- The input of CNN are image (three N by N if it color image and N by N if its black and white image)
- The weights are also two dimensional array

<img src="https://github.com/Make-School-Courses/DS-2.2-Deep-Learning/blob/master/Notebooks/Images/rgb_image.png?raw=true" width="500">

### The weights in CNN
The weights in CNN are called:
- Kernel
or
- Filter matrix
<img src="https://github.com/Make-School-Courses/DS-2.2-Deep-Learning/blob/master/Notebooks/Images/kernel_image.png?raw=true" width="600">

### Stride
- To define a CNN, we should specify the horizontal and vertical movement steps
- what is the output size with stride = 1 and stride =2?
<img src="https://github.com/Make-School-Courses/DS-2.2-Deep-Learning/blob/master/Notebooks/Images/stride_1.png?raw=true" width="700">
<img src="https://github.com/Make-School-Courses/DS-2.2-Deep-Learning/blob/master/Notebooks/Images/stride_2.png?raw=true" width="700">

- output_size = (input_size - filter_size)/stride + 1
- Stride visualization: http://deeplearning.stanford.edu/wiki/index.php/Feature_extraction_using_convolution

`model.add(Conv2D(32, kernel_size=(3, 3), strides=(1, 1), activation='relu', input_shape=input_shape))`

### Max Pooling vs Average Pooling
- Max pooling: take the maximum element from each window of a certain size
<img src="https://github.com/Make-School-Courses/DS-2.2-Deep-Learning/blob/master/Notebooks/Images/maxpooling.png?raw=true" width="700">

### Flatten Layer
- After feature extraction that is done by multiple Convolutional layers, we use flatten layer to add MLP after convolutional layers in order to do classification task
- This one is simple--it's just Keras's version of numpy.reshape. This reshapes n-dimensional arrays to a vector. This is necessary when moving from Conv2D layers, which expect 2-dimensional arrays as inputs, to Dense layers, which expect 1-dimension vectors as inputs. As a concrete example, a Flatten layer given a 28 x 28 array as input would output a vector of the shape (784, 1)

### Visualize the whole NN: CNN + MLP
<img src="https://github.com/Make-School-Courses/DS-2.2-Deep-Learning/blob/master/Notebooks/Images/CNN.png?raw=true" width="700">

### Activity: Obtain the number of parameters for the following CNN
By default, the strides = (1, 1)

In [26]:
from __future__ import print_function
import keras

from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D


model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=(28, 28, 1)))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
# model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
# model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

print(model.summary())

Model: "sequential_18"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_17 (Conv2D)           (None, 26, 26, 32)        320       
_________________________________________________________________
conv2d_18 (Conv2D)           (None, 24, 24, 64)        18496     
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 12, 12, 64)        0         
_________________________________________________________________
flatten_6 (Flatten)          (None, 9216)              0         
_________________________________________________________________
dense_13 (Dense)             (None, 128)               1179776   
_________________________________________________________________
dense_14 (Dense)             (None, 10)                1290      
Total params: 1,199,882
Trainable params: 1,199,882
Non-trainable params: 0
___________________________________________

#### Solution
output_size = (28 - 3)/1 + 1 = 26

output_size = (26 - 3)/1 + 1 = 24

The parameters for the first Conv2D = 32 x 9 + 32 = 320

The parameters for the second Conv2D = 64 x 32 x 9 + 64 = 18496

The shape for flatten is: 12 x 12 x 64 = 9216

The parameters for dense_1 = 9216 x 128 + 128 = 1179776

The parameters for dense_2 = 128 x 10 + 10 = 1290

### Data Preparation for CNN
Suppose we want to feed a 4 by 4 image to a CNN network, how we should reshape the data?

In [27]:
from keras.layers import Input, Dense, Conv2D, MaxPooling2D, UpSampling2D
from keras.models import Model
from keras import backend as K
import numpy as np

input_img = Input(shape=(4, 4, 1))  # adapt this if using `channels_first` image data format

x = Conv2D(2, (2, 2), activation='relu')(input_img)
y = Conv2D(3, (2, 2), activation='relu')(x)
model = Model(input_img, y)
# cnv_ml_1 = Model(input_img, x)

data = np.array([[5, 12, 1, 8], [2, 10, 3, 6], [4, 7, 9, 1], [5, 7, 5, 6]])
data = data.reshape(1, 4, 4, 1)
print(model.predict(data))
print('M :')
print(model.predict(data).reshape(3, 2, 2))
print(model.summary())
from keras.datasets import mnist

img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

print(x_train[0].shape)
print(x_train[1].shape)

[[[[ 4.181149  17.891964   0.       ]
   [ 5.3029346 15.295205   0.       ]]

  [[ 3.5036564 18.106941   0.       ]
   [ 4.926444  15.246883   0.       ]]]]
M :
[[[ 4.181149  17.891964 ]
  [ 0.         5.3029346]]

 [[15.295205   0.       ]
  [ 3.5036564 18.106941 ]]

 [[ 0.         4.926444 ]
  [15.246883   0.       ]]]
Model: "functional_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 4, 4, 1)]         0         
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 3, 3, 2)           10        
_________________________________________________________________
conv2d_20 (Conv2D)           (None, 2, 2, 3)           27        
Total params: 37
Trainable params: 37
Non-trainable params: 0
_________________________________________________________________
None
(28, 28, 1)
(28, 28, 1)


<a name="day10"></a> 
# Day 10: Recurrent Neural Network

